src/kakarot/gas.cairo

```diff
@@ -137,25 +137,37 @@ namespace Gas {
         words_len: felt, offset_1: Uint256*, size_1: Uint256*, offset_2: Uint256*, size_2: Uint256*
     ) -> model.MemoryExpansion {
         alloc_locals;
+
         let (is_zero_1) = uint256_eq([size_1], Uint256(0, 0));
         let (is_zero_2) = uint256_eq([size_2], Uint256(0, 0));
-        let is_zero = is_zero_1 * is_zero_2;
-        if (is_zero != FALSE) {
-            let expansion = model.MemoryExpansion(cost=0, new_words_len=words_len);
-            return expansion;
-        }
-        let max_expansion_is_2 = is_nn(offset_2.low + size_2.low - (offset_1.low + size_1.low));
-        let max_expansion = max_expansion_is_2 * (offset_2.low + size_2.low) + (
-            1 - max_expansion_is_2
-        ) * (offset_1.low + size_1.low);
+        tempvar both_zero = is_zero_1 * is_zero_2;
+        jmp no_expansion if both_zero != 0;
+
+        tempvar is_not_saturated = Helpers.is_zero(offset_1.high) * Helpers.is_zero(size_1.high) *
+            Helpers.is_zero(offset_2.high) * Helpers.is_zero(size_2.high);
+        tempvar is_saturated = 1 - is_not_saturated;
+        jmp expansion_cost_saturated if is_saturated != 0;
 
-        let low_expansion = calculate_gas_extend_memory(words_len, max_expansion);
-        let expansion_cost = low_expansion.cost + (
-            offset_1.high + size_1.high + offset_2.high + size_2.high
-        ) * Gas.MEMORY_COST_U128;
+        let max_offset_1 = (1 - is_zero_1) * (offset_1.low + size_1.low);
+        let max_offset_2 = (1 - is_zero_2) * (offset_2.low + size_2.low);
+        let max_expansion_is_2 = is_le_felt(max_offset_1, max_offset_2);
+        let max_offset = max_offset_1 * (1 - max_expansion_is_2) + max_offset_2 *
+            max_expansion_is_2;
+        let expansion = calculate_gas_extend_memory(words_len, max_offset);
+        let expansion = model.MemoryExpansion(
+            cost=expansion.cost, new_words_len=expansion.new_words_len
+        );
+        return expansion;
+
+        no_expansion:
+        let range_check_ptr = [fp - 8];
+        let expansion = model.MemoryExpansion(cost=0, new_words_len=words_len);
+        return expansion;
 
+        expansion_cost_saturated:
+        let range_check_ptr = [fp - 8];
         let expansion = model.MemoryExpansion(
-            cost=expansion_cost, new_words_len=low_expansion.new_words_len
+            cost=Gas.MEMORY_COST_U128, new_words_len=0x8000000000000000000000000000000
```

Comment:
Maybe add the same explanation as in `memory_expansion_cost_saturated` to have context on why their is a hardcoded value and what will happen in this case ?
https://github.com/kkrt-labs/kakarot/blob/7bc70ad9965ecb69d5242957594339975a0b3a8c/src/kakarot/gas.cairo#L114